### ISSUE TITLE
amp-action-macro Issue

### ISSUE THREAD

**Issue by ali33** [ampproject/amphtml#25735]
https://github.com/ampproject/amphtml/blob/master/examples/amp-action-macro.html

![1](https://user-images.githubusercontent.com/9099754/69415928-16d3b100-0d48-11ea-8fe1-8a08d8752ae1.PNG)
![2](https://user-images.githubusercontent.com/9099754/69415954-20f5af80-0d48-11ea-9674-153c20cefa27.PNG)
The element did not specify a layout attribute. Check https://amp.dev/documentation/guides-and-tutorials/develop/style_and_layout/control_layout and the respective element documentation for details.
vf @ error.js:195
(anonymous) @ error.js:128
f.assert @ log.js:428
D @ log.js:988
Fi.b.mg @ custom-element.js:430
Fi.b.jd @ custom-element.js:415
Fi.b.$f @ custom-element.js:910
Fi.b.upgrade @ custom-element.js:389
Ii @ custom-element-registry.js:89
dm @ custom-element-registry.js:73
cm @ extensions-impl.js:347
(anonymous) @ extensions-impl.js:323
f.addDocFactory @ extensions-impl.js:408
f.addElement @ extensions-impl.js:322
(anonymous) @ amp-action-macro.js:122
f @ amp-action-macro.js:121
f.registerExtension @ extensions-impl.js:154
(anonymous) @ runtime.js:205
Promise.then (async)
amp-action-macro @ runtime.js:201
wi @ chunk.js:192
si.Ze @ chunk.js:410
(anonymous) @ chunk.js:461
Promise.then (async)
zi @ chunk.js:460
si.ra @ chunk.js:479
si.runForStartup @ chunk.js:369
ti @ chunk.js:69
im @ runtime.js:333
b @ runtime.js:218
(anonymous) @ runtime.js:260
(anonymous) @ timer-impl.js:83
setTimeout (async)
f.delay @ timer-impl.js:89
km @ runtime.js:998
gm @ runtime.js:242
lm @ runtime.js:344
(anonymous) @ amp.js:118
wi @ chunk.js:192
si.Ze @ chunk.js:410
(anonymous) @ chunk.js:461
Promise.then (async)
zi @ chunk.js:460
si.ra @ chunk.js:479
(anonymous) @ chunk.js:436
Promise.then (async)
si.Ze @ chunk.js:416
(anonymous) @ chunk.js:461
Promise.then (async)
zi @ chunk.js:460
si.ra @ chunk.js:479
(anonymous) @ chunk.js:436
Promise.then (async)
si.Ze @ chunk.js:416
(anonymous) @ chunk.js:461
Promise.then (async)
zi @ chunk.js:460
si.ra @ chunk.js:479
si.runForStartup @ chunk.js:369
ti @ chunk.js:69
(anonymous) @ amp.js:91
(anonymous) @ amp.js:66.)

**Comment by caroqliu** [comment]
Hi @ali33, thanks for filing this issue! Want to clarify on the issue -- is the error you are seeing in the console coming from using a code snippet in the example markup you linked?

**Comment by ali33** [comment]
> Hi @ali33, thanks for filing this issue! Want to clarify on the issue -- is the error you are seeing in the console coming from using a code snippet in the example markup you linked?

Hi carqliu,
When I use amp-action-macro, the console write error message 'need layout attribute' but when I use layout attribute it write other error message.

**Comment by westonruter** [comment]
@ali33 Could you please share the specific AMP HTML markup you are using to cause the issue?

**Comment by ali33** [comment]
> @ali33 Could you please share the specific AMP HTML markup you are using to cause the issue?
@westonruter  You can use the example on github or access the copy here https://meta.vn/a/amp/amp-action-macro.htm. And then open console tab in chrome, You'll see the error message:
`The element did not specify a layout attribute. Check https://amp.dev/documentation/guides-and-tutorials/develop/style_and_layout/control_layout and the respective element documentation for details.`

**Comment by westonruter** [comment]
@ali33 Thanks. Yes, I can see the same thing. Using this markup:

```html
<amp-action-macro id="scrollable-carousel-macro-args" execute="scrollable-carousel.goToSlide(index=x)" arguments="x">
</amp-action-macro>
```

I get an error in the console:

> The element did not specify a layout attribute. Check https://amp.dev/documentation/guides-and-tutorials/develop/style_and_layout/control_layout and the respective element documentation for details.

If I then supply a `layout` attribute:

```html
<amp-action-macro layout="nodisplay" id="scrollable-carousel-macro-args" execute="scrollable-carousel.goToSlide(index=x)" arguments="x">
</amp-action-macro>
```

Then the AMP validator complains:

> The attribute 'layout' may not appear in tag 'amp-action-macro'.

This would be as expected because the validator spec doesn't indicate the `layout` attribute is allowed:

https://github.com/ampproject/amphtml/blob/63fa1d05386d20ff6135a42000cf9d3036b253cd/extensions/amp-action-macro/validator-amp-action-macro.protoascii#L32-L39

It makes sense to me that the element doesn't allow the `layout` attribute because this element should always be `nodisplay`. So the problem to me that the problem is that the `amp-action-macro` isn't defining a default value for the  `layout_` property, which is then causing `assertLayout_()` method to fail:

https://github.com/ampproject/amphtml/blob/63fa1d05386d20ff6135a42000cf9d3036b253cd/src/custom-element.js#L424-L439

But this doesn't make sense to me either because a default value for `layout_` _is_ being set:

https://github.com/ampproject/amphtml/blob/63fa1d05386d20ff6135a42000cf9d3036b253cd/src/custom-element.js#L183-L184

This is happening in the `createdCallback` whereas the `assertLayout_` call is happening in ` completeUpgrade_`.

So apparently `completeUpgrade_` is being called before `createdCallback`, and this is causing a problem.

**Comment by westonruter** [comment]
cc @alabiaga as code owner (for the `amp-action-macro`)

**Comment by alabiaga** [comment]
@westonruter thanks Weston, I'll take a look later today

**Comment by alabiaga** [comment]
So this is an issue with all non visible, no layout supported components. Issue is pointed out by @westonruter in https://github.com/ampproject/amphtml/issues/25735#issuecomment-561530969. Components get around it by just overriding isLayoutSupported and returning true. The fix in #26280 will follow suit and I will defer to the runtime folks for the proper fix. Thanks
